src/AbstractMigrator.php

```diff
@@ -0,0 +1,395 @@
+     *
+     * Overridden to:
+     * * pass different arguments to the callback progress function
+     * * allow preparation of SQL query strings before execution
+     * * contain a portion of the fix for the SQL file execution bug See {@link getMigrationSqlQuery()}.
+     * * throw specific exceptions
+     *
+     * @param int  $upVersion
+     * @param int  $increment Can accept 1 for UP or -1 for down
+     * @param bool $force
+     *
+     * @throws DatabaseIsIncompleteException
+     */
+    protected function migrate($upVersion, $increment, $force)
+    {
+        $versionInfo    = $this->getCurrentVersion();
+        $currentVersion = intval($versionInfo['version']);
+
+        if (strpos($versionInfo['status'], 'partial') !== false && !$force) {
+            throw new DatabaseIsIncompleteException(
+                'Database was not fully updated - use the "force" argument to ignore this error'
+            );
+        }
+
+        while (
+            $this->canContinue($currentVersion, $upVersion, $increment)
+            &&
+            $rawSql = $this->_getMigrationSqlQuery($currentVersion, $increment)
+        ) {
+            $nextVersion = $currentVersion + $increment;
+
+            $preparedSql = $this->_prepareSql($rawSql);
+
+            $this->getDbCommand()->setVersion($nextVersion, 'partial ' . ($increment > 0 ? 'up' : 'down'));
+            $this->getDbCommand()->executeSql($preparedSql);
+            $this->getDbCommand()->setVersion($nextVersion, 'complete');
+
+            $currentVersion = $nextVersion;
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Overridden to contain a portion of the fix for the SQL file execution bug. See {@link getMigrationSqlQuery()}.
+     * This method now properly checks if migration can be done from the current version to another version,
+     * including/excluding the current version depending on the increment.
+     *
+     * This is achieved by calculating the difference between the versions and multiplying by the increment. This result
+     * is called "delta", and it allows the detection of a discrepancy between the given versions and the sign of the
+     * increment by swapping the sign of the difference. If delta is zero or smaller, then the given versions and the
+     * increment are not consistent.
+     *
+     * Example:
+     * 1. If the up version is greater than the current version, the increment should be positive, their difference will
+     *    be positive and so delta will be positive.
+     * 2. If the up version is smaller than the current version, the increment should be negative, their difference will
+     *    be negative and so delta will be positive.
+     *
+     * @since [*next-version*]
+     */
+    protected function canContinue($currentVersion, $upVersion, $increment)
+    {
+        // Difference between versions
+        $delta = (intval($upVersion) - intval($currentVersion)) * $increment;
```

Comment:
I can't really understand how it is possible to compare versions as integers. What is the format of the version? Also, consider using `_normalizeInt()`.

Reply:
It's a simple integer, not like a version string separated by periods.

Will change to use `_normalizeInt()`.